post.md

```diff
@@ -0,0 +1,411 @@
+---
+post_title: Basics of x86 assembly (AT&T syntax)
+author: Bartek Pastudzki
+post_excerpt: "This article is meant to preasent the most important ideas behind x86 assembly just to show how does it work, what are its limitations."
+layout: post
+published: true
+---
+
+Nowadays, application of programming in assembly language is
+very small. Writing production code in assembly may be found
+in the most demanding tasks in embedded. Even in modern firmware
+(coreboot, EDK2) most of the code is written in C. Honesly,
+that's quite undestandable, assembly code isn't easy to read
+nor write and some state that it's no longer needed to be known.
+In my opinion, that's far from truth.
+
+Not only it's heavily used in reverse engineering, very important
+especially in software/firmware security, its design impact more
+or less code in virtualy any programming language. Let's set an
+example. Consider those two lines of C code:
+
+```c
+    char buffer[size];  // Since C99, we can do that with variable!!!
+    char *buffer2 = malloc(size);
+```
+
+Even C's syntax (considered low level nowadays) create an impression
+that they are more or less the same, except that `buffer` is `const char*`
+and `buffer2` is `char*` and the range: `buffer` is freed automaticaly at
+the and of current block. However someone with some knowledge of internls
+know that first line has virtually no overhead (because all local variables
+may be allocated and freed at the same time, it takes 2 instructions), while
+second one involves kernel activity to map needed space, needs to update heap
+structures, etc. Except that `malloc()` often happen to introduce memory,
+on the other hand first method may cause stack overflow in some cases.
+
+This is just one of many examples where low level details (assembly, network
+stack, devices, etc.) condition higher level code execution. That's why any
+senior software developer should know them. And of course any security
+specialist. Except that it's essentional to anyone interested in Reverse
+Enineering.
```

Comment:
This is just one of many examples where low-level details (assembly, network stack, devices, etc.) condition higher level code execution. That's why any senior software developer should know them and, of course, any security specialist. Except that, it's essential to anyone interested in Reverse Engineering.